### PR TITLE
论文复现34

### PR DESCRIPTION
**Pull request by marshall-dteach**

### CODE REVIEW

File: paddleseg/cvlibs/config.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import math
```
Comment (shiyutang) [inline paddleseg/cvlibs/config.py:15]: 没有用到需要删去

File: configs/psa/psa_hrnetv2_psa_cityscapes_1024x2048_520k.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
+_base_: '../_base_/cityscapes.yml'
```
Comment (shiyutang) [inline configs/psa/psa_hrnetv2_psa_cityscapes_1024x2048_520k.yml:1]: 没有使用到base config里的数据可以不用继承，同时 config的顺序可以调整为：
batch，trainset，valset，model， optimizer， lr——scheduler，loss

File: paddleseg/cvlibs/config.py [hunk 3]
```diff
@@ -166,6 +167,11 @@ def lr_scheduler(self) -> paddle.optimizer.lr.LRScheduler:
             params.setdefault('decay_steps', self.iters)
             params.setdefault('end_lr', 0)
             params.setdefault('power', 0.9)
+        elif lr_type == 'LambdaDecay':
+            params.pop('end_lr')
+            params.pop('power')
```
Comment (shiyutang) [inline paddleseg/cvlibs/config.py:172]: 为什么要删去这两个

File: paddleseg/models/backbones/__init__.py [hunk 4]
```diff
@@ -23,4 +23,5 @@
 from .stdcnet import *
 from .lite_hrnet import *
 from .shufflenetv2 import *
-from .ghostnet import *
\ No newline at end of file
+from .ghostnet import *
```
Comment (shiyutang) [inline paddleseg/models/backbones/__init__.py]: 不要修改这个

File: paddleseg/models/backbones/hrnetv2_psa.py [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+from paddleseg.cvlibs import manager
+from paddleseg.cvlibs import param_init
+
+
+class PSA_s(nn.Layer):
```
Comment (shiyutang) [inline paddleseg/models/backbones/hrnetv2_psa.py:25]: 类命名采取驼峰命名方式，同时取名需要有意义

File: paddleseg/models/backbones/hrnetv2_psa.py [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+        return out
+
+    def forward(self, x):
+        context_channel = self.spatial_pool(x)
+        context_spatial = self.channel_pool(x)
+        out = context_spatial + context_channel
+        return out
+
+
+class BasicBlock(nn.Layer):
```
Comment (shiyutang) [inline paddleseg/models/backbones/hrnetv2_psa.py:134]: 这个和resnet的重名了

File: paddleseg/models/backbones/__init__.py [hunk 7]
```diff
@@ -23,4 +23,5 @@
...
+from .hrnetv2_psa import *
```
Comment (shiyutang) [inline paddleseg/models/backbones/__init__.py:27]: 只需要import  HRNETV2_PSA就可以

File: paddleseg/models/backbones/hrnetv2_psa.py [hunk 8]
```diff
@@ -0,0 +1,603 @@
...
+        if self.downsample is not None:
+            residual = self.downsample(x)
+
+        out = out + residual
+        out = self.relu(out)
+
+        return out
+
+
+class Bottleneck(nn.Layer):
```
Comment (shiyutang) [inline paddleseg/models/backbones/hrnetv2_psa.py:168]: 是否可以直接利用这里的bottleneckBlock：https://github.com/PaddlePaddle/PaddleSeg/blob/9a358c0bc1e5e845e5127e1b7ad67df8cd4f1087/paddleseg/models/backbones/resnet_vd.py#L77

File: paddleseg/models/backbones/hrnetv2_psa.py [hunk 9]
```diff
@@ -0,0 +1,603 @@
...
+                   'bias') and self.conv_q_left.bias is not None:
+            param_init.constant_init(self.conv_q_left.bias, value=0.0)
+        param_init.kaiming_normal_init(self.conv_v_left.weight)
+        if hasattr(self.conv_v_left,
+                   'bias') and self.conv_v_left.bias is not None:
+            param_init.constant_init(self.conv_v_left.bias, value=0.0)
+        self.conv_q_right.inited = True
+        self.conv_v_right.inited = True
+        self.conv_q_left.inited = True
+        self.conv_v_left.inited = True
```
Comment (shiyutang) [inline paddleseg/models/backbones/hrnetv2_psa.py:95]: 这四个重复的可以用循环写起来，另外为什么self。conv_up不需要init？

File: paddleseg/models/layers/layer_libs.py [hunk 10]
```diff
@@ -350,3 +350,32 @@ def forward(self, x):
...
+            out_channels,
+            out_channels=out_channels,
+            kernel_size=1,
+            padding=padding,
+            **kwargs)
+
+    def forward(self, x):
+        x = self.conv_bn_relu_1(x)
+        x = self.conv_bn_relu_2(x)
+        return x
```
Comment (shiyutang) [inline paddleseg/models/layers/layer_libs.py:381]: 对于不常用的模块，可以直接放在models/psa.py中

File: test_tipc/configs/psa/psa_cityscapes_1024x2048_520k.yml [hunk 11]
```diff
@@ -0,0 +1,20 @@
+batch_size: 1
+iters: 520000
+
+model:
+    type: MscaleOCR
```
Comment (shiyutang) [inline test_tipc/configs/psa/psa_cityscapes_1024x2048_520k.yml:5]: 这个是什么？

File: test_tipc/prepare.sh [hunk 12]
```diff
@@ -91,6 +91,8 @@ if [ ${MODE} = "whole_infer" ] || [ ${MODE} = "klquant_whole_infer" ]; then
         wget -nc -P $model_path https://paddleseg.bj.bcebos.com/dygraph/cityscapes/pp_liteseg_stdc2_cityscapes_1024x512_scale1.0_160k/model.pdparams
     elif [ ${model_name} == "ddrnet" ];then
         wget -nc -P $model_path https://bj.bcebos.com/paddleseg/dygraph/cityscapes/ddrnet23_cityscapes_1024x1024_120k/model.pdparams
+    elif [ ${model_name} == "psa" ];then
+        wget -nc -P $model_path https://bj.bcebos.com/paddleseg/dygraph/cityscapes/psa_hrnetv2_psa_cityscapes_1024x2048_520k/model.pdparams
```
Comment (shiyutang) [inline test_tipc/prepare.sh:95]: 这个模型已经上传了么？

File: paddleseg/models/psa.py [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+        context = context.transpose((0, 2, 1))
+        context = context.reshape((batch_size, self.key_channels,
+                                   *paddle.shape(x).numpy()[2:]))
+        context = self.f_up(context)
+        if self.scale > 1:
+            context = F.interpolate(context, size=(h, w), mode='bilinear')
+        return context
+
+
+class SpatialOCR_Module(nn.Layer):
```
Comment (shiyutang) [inline paddleseg/models/psa.py:140]: 类都需要使用驼峰命名的规范

File: paddleseg/models/psa.py [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+    def __init__(self,
+                 in_channels,
+                 key_channels,
+                 out_channels,
+                 scale=1,
+                 dropout=0.1):
+        super().__init__()
+        self.object_context_block = ObjectAttentionBlock(in_channels,
```
Comment (shiyutang) [inline paddleseg/models/psa.py:148]: 只有这一处用到了就不用写成一个类，不然太乱了

File: paddleseg/models/psa.py [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+        return logit_list
+
+    def init_weight(self):
+        if self.pretrained is not None:
+            utils.load_entire_model(self, self.pretrained)
+
+    def forward(self, inputs):
+        if [0.5, 1.0, 2.0] and not self.training:
+            return self.nscale_forward(inputs, [0.5, 1.0, 2.0])
+        return self.two_scale_forward(inputs)
```
Comment (shiyutang) [inline paddleseg/models/psa.py:357]: 目前模型写的太乱了，最好可以变成下面这样，把只用到了一次的模块变成类的静态方法，同时简单分为backbone，和head两个部分，在head里面再去做选择等。
image

File: paddleseg/models/psa.py [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+        for sublayer in self.sublayers():
+            if isinstance(sublayer, nn.Conv2D):
+                param_init.normal_init(sublayer.weight, std=0.001)
+            elif isinstance(sublayer, (nn.BatchNorm, nn.SyncBatchNorm)):
+                param_init.constant_init(sublayer.weight, value=1.0)
+                param_init.constant_init(sublayer.bias, value=0.0)
+
+
+@manager.MODELS.add_component
+class PSA(nn.Layer):
```
Comment (shiyutang) [inline paddleseg/models/psa.py:227]: 模型名字要完整，PSANet

**Review by shiyutang — CHANGES_REQUESTED** [review]
整体结构有了，但是建议很多细节需要完善，包含类的命名，网络组织结构，类的组织方式等

File: paddleseg/cvlibs/config.py (hunk 1 above)
Comment (marshall-dteach) [reply paddleseg/cvlibs/config.py:15]: 您好，这个虽然看上去没用，但是在lr_lambda中加载的时候会用到math.ceil取整

File: configs/psa/psa_hrnetv2_psa_cityscapes_1024x2048_520k.yml (hunk 2 above)
Comment (marshall-dteach) [reply configs/psa/psa_hrnetv2_psa_cityscapes_1024x2048_520k.yml:1]: 收到

File: paddleseg/cvlibs/config.py (hunk 3 above)
Comment (marshall-dteach) [reply paddleseg/cvlibs/config.py:172]: 因为lambdadency里没有end_lr和power参数，如果不继承base，这里可以修改，免去删掉这两个

File: paddleseg/models/backbones/__init__.py (hunk 4 above)
Comment (marshall-dteach) [reply paddleseg/models/backbones/__init__.py]: 这里是发生conflict了，等我把问题都弄懂之后，重新fork一下

File: paddleseg/models/backbones/hrnetv2_psa.py (hunk 5 above)
Comment (marshall-dteach) [reply paddleseg/models/backbones/hrnetv2_psa.py:25]: 收到

File: paddleseg/models/backbones/hrnetv2_psa.py (hunk 6 above)
Comment (marshall-dteach) [reply paddleseg/models/backbones/hrnetv2_psa.py:134]: 收到，我会修改的

File: paddleseg/models/psa.py (hunk 16 above)
Comment (marshall-dteach) [reply paddleseg/models/psa.py:227]: 收到

File: paddleseg/models/psa.py (hunk 15 above)
Comment (marshall-dteach) [reply paddleseg/models/psa.py:357]: 好的，因为好多地方不是很懂，不太敢动，但我会尽量去改

File: paddleseg/models/psa.py (hunk 14 above)
Comment (marshall-dteach) [reply paddleseg/models/psa.py:148]: 收到

File: paddleseg/models/psa.py (hunk 13 above)
Comment (marshall-dteach) [reply paddleseg/models/psa.py:140]: 收到

File: test_tipc/prepare.sh (hunk 12 above)
Comment (marshall-dteach) [reply test_tipc/prepare.sh:95]: 静态模型大于64M，直接加载没法加载，因此没上传，只修改了相应的文件

Comment (marshall-dteach) [reply test_tipc/prepare.sh:95]: 动态模型，我放在了百度网盘里，您需要的话分享给您

File: test_tipc/configs/psa/psa_cityscapes_1024x2048_520k.yml (hunk 11 above)
Comment (marshall-dteach) [reply test_tipc/configs/psa/psa_cityscapes_1024x2048_520k.yml:5]: 这个文件好像是之前的，我记得删掉了，但不知为什么又给我push上来了，我删掉重新修改一下

File: paddleseg/models/layers/layer_libs.py (hunk 10 above)
Comment (marshall-dteach) [reply paddleseg/models/layers/layer_libs.py:381]: 收到

File: paddleseg/models/backbones/hrnetv2_psa.py (hunk 9 above)
Comment (marshall-dteach) [reply paddleseg/models/backbones/hrnetv2_psa.py:95]: 我改一下，这个conv_up的初始化我记得被注释了，我也好奇来着，我再看一下

File: paddleseg/models/backbones/hrnetv2_psa.py (hunk 8 above)
Comment (marshall-dteach) [reply paddleseg/models/backbones/hrnetv2_psa.py:168]: 好的，我看一下

File: paddleseg/models/backbones/__init__.py (hunk 7 above)
Comment (marshall-dteach) [reply paddleseg/models/backbones/__init__.py:27]: 收到

File: paddleseg/models/backbones/hrnetv2_psa.py (hunk 9 above)
Comment (marshall-dteach) [reply paddleseg/models/backbones/hrnetv2_psa.py:95]: 也可能是我改的时候误删了

File: test_tipc/prepare.sh (hunk 12 above)
Comment (shiyutang) [reply test_tipc/prepare.sh:95]: 没有上传这里就不用增加，链接是上传之后才有的

**Comment by shiyutang** [comment]
修改完之后需要重新 re-request review或者微信提醒我 review， 从而进一步推进。

File: paddleseg/cvlibs/config.py (hunk 1 above)
Comment (shiyutang) [reply paddleseg/cvlibs/config.py:15]: 好的

File: paddleseg/models/psa.py (hunk 15 above)
Comment (marshall-dteach) [reply paddleseg/models/psa.py:357]: 您好，我又看了下源代码，我的理解是涉及到多尺度，因此logits_list的内容涉及到相应的变化而并不是直接插值来的，我用_fwd函数区分了backbone和head两部分，如果直接改成实例代码类型forward中可能会更乱
